### PR TITLE
[WIP] New Array.filter - better memory usage

### PR DESCRIPTION
**Pull request by jackmott**
The motivation here is to keep the same overall speed as the new Array.filter from a few days ago, without having the greatly increased memory use in scenarios when most of the array elements are filtered.

This implementation is similar to the original one which used a ResizeArray, but takes advantage of the fact that the growth of the array is bounded by array.Length so that we never allocate ~2*array.Length.  Additionally, two fast paths exist. If all elements are filtered, nothing is allocated and empty is returned directly. If no elements are filtered, the Array.Copy at the end is skipped, because our resized array is exactly the right size in that case, also improving speed and reducing memory usage.

Overall, this has similar runtime as the current Array.filter with about 25% to 35% less allocation. The benchmark below shows the results for the current, original, and new filter, when 0, 10, 25, 75, and 90, and 100% of elements remain after filtering.

The downside to this approach is it is more complex code.

If you like this change, the same logic would be applied to choose.

```ini

Host Process Environment Information:
BenchmarkDotNet=v0.9.8.0
OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i7-4712HQ CPU 2.30GHz, ProcessorCount=8
Frequency=2240910 ticks, Resolution=446.2473 ns, Timer=TSC
CLR=MS.NET 4.0.30319.42000, Arch=64-bit RELEASE [RyuJIT]
GC=Concurrent Workstation
JitModules=clrjit-v4.6.1590.0

Type=CoreBenchmark  Mode=Throughput  Platform=X64  
Jit=RyuJit  GarbageCollection=Concurrent Workstation  

```
  Method | Length | Pct |        Median |     StdDev | Scaled |  Gen 0 | Gen 1 |    Gen 2 | Bytes Allocated/Op |
-------- |------- |---- |-------------- |----------- |------- |------- |------ |--------- |------------------- |
     **Old** | **100000** |   **0** |   **257.9468 us** |  **3.1356 us** |   **1.00** |      **-** |     **-** |        **-** |             **437.56** |
 Current | 100000 |   0 |   207.4557 us |  7.2287 us |   0.80 |      - |     - |   569.59 |         168,274.25 |
    New3 | 100000 |   0 |   192.8935 us |  4.2597 us |   0.75 |      - |     - |        - |             223.45 |
     **Old** | **100000** |  **10** |   **388.3678 us** |  **6.3276 us** |   **1.00** | **231.46** |  **9.50** |        **-** |          **87,711.42** |
 Current | 100000 |  10 |   323.8600 us |  7.0283 us |   0.83 |      - |     - |   550.66 |         181,880.46 |
    New3 | 100000 |  10 |   371.5841 us |  3.5846 us |   0.96 |      - |     - |   150.93 |          68,134.81 |
     **Old** | **100000** |  **25** |   **649.2728 us** | **15.1429 us** |   **1.00** |      **-** |     **-** |   **257.15** |         **132,033.02** |
 Current | 100000 |  25 |   552.9257 us |  8.8261 us |   0.85 |      - |     - |   615.26 |         224,125.63 |
    New3 | 100000 |  25 |   540.5000 us |  9.4800 us |   0.83 |      - |     - |   270.17 |          83,741.20 |
     **Old** | **100000** |  **50** | **1,021.8182 us** | **19.1519 us** |   **1.00** |      **-** |     **-** |   **824.00** |         **320,059.78** |
 Current | 100000 |  50 |   756.3990 us |  9.5827 us |   0.74 |      - |     - |   757.50 |         229,800.24 |
    New3 | 100000 |  50 |   803.6203 us | 26.9094 us |   0.79 |      - |     - |   676.62 |         204,549.69 |
     **Old** | **100000** |  **75** | **1,062.5958 us** | **25.0791 us** |   **1.00** | **172.00** |  **1.95** | **1,492.28** |         **563,836.73** |
 Current | 100000 |  75 |   484.5561 us | 12.2847 us |   0.46 |      - |     - |   867.11 |         278,742.70 |
    New3 | 100000 |  75 |   658.7146 us | 11.9907 us |   0.62 |      - |     - |   866.10 |         265,235.53 |
     **Old** | **100000** |  **90** |   **960.1716 us** |  **8.5740 us** |   **1.00** | **157.34** | **16.61** | **1,492.28** |         **593,153.20** |
 Current | 100000 |  90 |   429.5871 us | 35.2049 us |   0.45 |      - |     - |   961.71 |         331,770.50 |
    New3 | 100000 |  90 |   502.8985 us | 13.9179 us |   0.52 |      - |     - |   991.75 |         308,047.74 |
     **Old** | **100000** | **100** |   **624.1714 us** | **15.1284 us** |   **1.00** |      **-** |     **-** | **1,850.08** |         **628,711.50** |
 Current | 100000 | 100 |   399.4854 us | 11.1079 us |   0.64 |      - |     - |   944.05 |         279,019.83 |
    New3 | 100000 | 100 |   465.3832 us |  9.6660 us |   0.75 |      - |     - | 1,113.40 |         328,896.22 |

### CODE REVIEW

**Comment by PatrickMcDonald** [comment]
As f can now be evaluated multiple times per element, this could have worse performance if f is slow. Also, if f has side effects this implementation could have different behavior from previous.

**Comment by jackmott** [comment]
It should not be the case that f is evaluated multiple times per element. The initial loop stops at the first element that returns true, and does not re-evaluate it, unless I have made an error.

**Comment by PatrickMcDonald** [comment]
Apologies, my mistake

**Comment by jackmott** [comment]
So this changes builds successfully, but when I try to run tests on it I get:

> C:\projects\visualfsharp\tests>RunTests.cmd release coreunit
C:\projects\visualfsharp\tests\..\release\net40\bin
"C:\projects\visualfsharp\tests\..\release\net40\bin\Fsi.exe" --exec --nologo "C:\projects\visualfsharp\tests\Convert-NUnit2Args-to-NUnit3Where.fsx" -- "" "NOOPEN,Expensive"


>error FS0193: internal error: Object reference not set to an instance of an object.
Error converting args to nunit 3 test selection language, the nunit3-console --where argument


Any ideas?

**Comment by asik** [comment]
Are you sure this works? Just copy-pasted it in F# interactive, replaced Array.zeroCreateUnchecked and Array.subUnchecked with their non-unchecked equivalents to make it compile, and I get:

```
filter (fun a -> a % 2 = 0) [|1;2;3;4;5;6;7;8;9|];;
val it : int [] = [|0; 4; 6; 8|]

Array.filter (fun a -> a % 2 = 0) [|1;2;3;4;5;6;7;8;9|];;
val it : int [] = [|2; 4; 6; 8|]
```

**Comment by jackmott** [comment]
@asik - been struggling with the build today so no, let me revisit.

**Comment by jackmott** [comment]
Ok fix is in, and I've rerun the same benchmark, and updated the original post with it.

**Comment by asik** [comment]
Ok so let me understand what this is doing :

> STEP 1 (0 allocations)
>  - Find the first element in the input for which p is true
>  - If there's none, return empty.
> 
> STEP 2 (Total: 1 allocation + probably 1 for Array.sub if we're done)
>  - Else allocate an output array of about a quarter the number of remaining elements.
>  - Put that first element it.
>  - Keep using this array to put elements for which p is true.
>  - If we're done, just Array.sub that (or avoid the Array.sub entirely if it happens to have the right size)
> 
> STEP 3 (Total: 2 allocations + probably 1 for Array.sub)
>  - Else allocate a second output array of the number of remaining elements + the size of the first array
>  - Copy all the elements of the first array into that new array
>  - And keep going until we're done.
>  - Array.sub that or avoid it if possible.

We can avoid the extraneous copy and reduce memory pressure in step 3 like so:

> STEP 3 (Total: 3 allocations)
>  - Else allocate a second output array of only the number of remaining elements
>  - Keep going until we're done.
>  - Allocate a result array of [count] elements and copy the first and second output arrays in it.
>  - Return that array

This is suboptimal in cases where Array.sub could coincidentally be avoided, but those should be relatively rare. In the general case I expect this to be faster. Example code:

```
        if i < array.Length then                            
            let res2 = Array.zeroCreate (array.Length-i)     
            while i < array.Length do
                element <- array.[i]                                
                if f element then                    
                    res2.[count - res.Length] <- element
                    count <- count + 1                            
                i <- i + 1

            let final = Array.zeroCreate (count)
            Array.Copy(res, final, res.Length)
            Array.Copy(res2, 0, final, res.Length, count - res.Length)
            final
        else
            Array.sub res 0 count
```

More food for thought:

Speaking more generally, this algorithm is an optimization on the "naive" one submitted earlier because it tries to avoid allocating the full array and starts with an (I assume) empirically-based estimate of about a quarter of the size, and never allocates more than one extra array, if necessary, to do this.

The worst case for this algorithm would be a large array of large structs where the first 25% of elements are positive and the last 75% are negative. In this case it will allocate a large amount of memory unnecessarily. But it's not doing much worse in that regard than the original algorithm.

What I'm wondering is if the partitioning technique used here could not be refined more. In particular, out there in nature, I suspect most Array.filter are done on relatively small arrays, and here the naive algorithm might still win just by virtue of only ever doing 2 allocations and generating tighter assembly. 

So here's a generalization of the above algorithm:
```
// Determine the optimal number of partitions based on the size in bytes of the input
// Allocate the first partition ('T[]), set it as current
// for i = 0 to array.Length - 1 do
//    if we ran out of space in the current partition, allocate the next one and set it as current
//    if f array.[i] then
//         currentPartition.[i] <- array.[i]
//         count <- count + 1
// let result = Array.zeroCreateUnchecked count
// copy the contents of each partition one after the other into result
// result
```

This requires empirical testing, but this optimal number of partitions is probably something like:
```
if arraySizeInBytes < 64 then 1
elif arraySizeInBytes < 1024 then 2
elif ...
else maxPartitions
```
And now this is as fast as the naive algorithm for small arrays, performs consistently even in the "poison" large-array-of-structs case, and never allocates a large amount of memory that we might not need in the end. However, it should get worse in terms of number of allocations as the size of the array grows, while still handily beating the original approach.

**Comment by jackmott** [comment]
Good ideas to test, I will try them out.

**Comment by jackmott** [comment]
@asik - Your first optimization was a clear win, even including the `nothing is filtered` in 1 out of 7 cases it was good for ~15% less memory use so the total memory use improvement is bumped up to ~25% now, or ~35% if you ignore the worst case of nothing being filtered. Running unit tests then will push it up.

Next up will be setting up a generalized version to experiment with your idea of changing the number of partitions by array length.  I don't know that overall results will be very sensitive to it based on other experiments I've been doing, but only one way to find out.

**Comment by asik** [comment]
Great! 

It should really be by array size in bytes, the point is to minimize the size of allocations. That's `array.Length * sizeof<T>` for structs, and `array.Length * sizeof<IntPtr>` for reference types.

I'm not sure it'll be any faster on average, but it should minimize memory pressure especially for large arrays of structs, which I find problematic in the current implementation; performance should be about the same as your new proposal.

I think your insight of using unequal partitions (25%-75%) is a very useful one, it uses the fact that the distribution of the function p is likely to stay constant over the input array to minimize allocations. Based on this I think a good starting point would consist of roughly doubling the size of the partition each time it grows, i.e.

1 - 100% (trivial case)
2 - (33% - 67%, or perhaps 25% - 75% as you found?)
3 - (14% - 29% - 57%)
4 - (7% - 13% - 27% - 53%)

**Comment by jackmott** [comment]
I did some experiments with the initial array size, 1/2, 1/4, 1/8, and it really didn't make much difference. It just trades speed for memory in small amounts depending on the structure of the input array. 1/4 seemed to be a slight winner overall.

**Comment by jackmott** [comment]
@asik Is there any way to tell if the array contains value or reference types efficiently?

**Comment by manofstick** [comment]
Hi guys,

Wondering if you had the chance to benchmark [this](https://github.com/Microsoft/visualfsharp/pull/1414#issuecomment-240924859) yet?

Don't have access to a machine for a bit so can tweak at the moment...

**Comment by jackmott** [comment]
@manofstick I did - memory use was good but runtime was slow.  You can do a similar approach using BitArray which makes for a lot less code, not sure how much overhead.  I had tried it earlier but it was also slow. I think the problem with the boolean approach is you end up with more branching in the loops.

**Comment by asik** [comment]
Actually, just sizeof<'T> for all types is ok. It returns the size of the reference for reference types.

**Comment by manofstick** [comment]
@jackmott 

Hmmm... Well the problem with BitArray is that it knows a lot less about what you are doing... The operations I'm doing are pretty trivial, i guess it loses out on the array access jit optimisations. I was getting pretty good times when I did run it where the filtered list returned a reasonable proportion of the results, depreciating when run with the 64-bit jit. Could you post a gist to your test suite?

**Comment by jackmott** [comment]
@manofstick  you can use this as a template: https://gist.github.com/jackmott/d160e7300ac8a1ceef393ed8f48bada6

Also be sure to test with arrays of random ints.  If you happened to be testing with a sorted list, branch prediction will always fly right through the if statements and perform really well.

**Comment by jackmott** [comment]
@asik  here is a prototype of the tuneable filter.  If you have ideas on how to simplify / improve let me know https://gist.github.com/jackmott/c75f7e871ded1a787810161be2384148

Currently partition size is constant, rather than doubling.  Not for any good reason, just keeping it simple to start.

**Comment by asik** [comment]
Well I fixed some bugs and optimized it to death, I feel almost dirty writing code like this.
https://gist.github.com/asik/66575daa9c09d0d8923728d9ed885407 
But based on some initial tests it seems slow compared to yours, what might be killing it is always having the double condition
`while count < currentChunk.Length && i < array.Length do`
while yours avoids this in step 3 by virtue of knowing this is the last array.

It would be a lot of work trying to fine-tune the partition sizes to minimize GC pressure and get a "good" time-space tradeoff.

**Comment by jackmott** [comment]
So doing some testing with your code,  but adding doubling growth, with 1 million ints, you can get some memory wins in some cases, but not all, and speed ends up ~20-30% worse in all cases. The big memory wins are when you filter away 99 to 95% of the items. Then it sorta alternates between slightly better or worse depending on how lucky you get.

I think what is currently in the PR may represent a pretty good compromise. It is somewhat self-tuning, like if you have a really sparse result, you will spin through a lot of the initial elements, and the first allocation will be 1/4 of the remaining possible elements, not just 1/4 of the array.

File: src/fsharp/FSharp.Core/array.fs [hunk 1]
```diff
@@ -483,19 +483,45 @@ namespace Microsoft.FSharp.Collections
...
-            for x in array do                 
-                if f x then 
-                    res.[count] <- x
-                    count <- count + 1
-            Array.subUnchecked 0 count res
-
+        let filter f (array: _[]) =         
+            checkNonNull "array" array                    
+            let mutable i = 0    
+            while i < array.Length && not (f array.[i]) do
```
Comment (KevinRansom) [inline src/fsharp/FSharp.Core/array.fs]: This is a really neat optimization, it means that filters that find nothing allocate nothing.  And filters that don't hit early in the iteration allocate fewer elements.  And it doesn't even do more work to do that.

I like this a lot.

Kevin

File: src/fsharp/FSharp.Core/array.fs [hunk 2]
```diff
@@ -483,19 +483,45 @@ namespace Microsoft.FSharp.Collections
...
+                i <- i + 1
+            
+            if i <> array.Length then                    
+                let mutable element = array.[i]
+                let chunk1 : 'T[] = Array.zeroCreateUnchecked (((array.Length-i) >>> 2) + 1)
```
Comment (KevinRansom) [inline src/fsharp/FSharp.Core/array.fs]: @jackmott 
I'm not sure about the chunking heuristic.  Just out of curiosity ... what would the results look like ... if you took the above optimisation and applied it to the existing algorithm?

Comment (jackmott) [inline src/fsharp/FSharp.Core/array.fs]: Sometimes a little faster, sometimes a little slower, depending on how much gets filtered.
Big memory penalty when many things get filtered, which also causes a speed penalty. It does restore the fast path for when nothing gets filtered.

```ini

Host Process Environment Information:
BenchmarkDotNet=v0.9.8.0
OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i7-4712HQ CPU 2.30GHz, ProcessorCount=8
Frequency=2240910 ticks, Resolution=446.2473 ns, Timer=TSC
CLR=MS.NET 4.0.30319.42000, Arch=64-bit RELEASE [RyuJIT]
GC=Concurrent Workstation
JitModules=clrjit-v4.6.1590.0

Type=CoreBenchmark  Mode=Throughput  Platform=X64  
Jit=RyuJit  GarbageCollection=Concurrent Workstation  

```
  Method | Length | Pct |      Median |     StdDev | Scaled | Gen 0 | Gen 1 |    Gen 2 | Bytes Allocated/Op |
-------- |------- |---- |------------ |----------- |------- |------ |------ |--------- |------------------- |
     **New** | **100000** |   **0** | **192.6800 us** |  **3.7269 us** |   **0.99** |     **-** |     **-** |        **-** |             **223.45** |
 OldPlus | 100000 |   0 | 194.2310 us |  5.4061 us |   1.00 |     - |     - |        - |             222.20 |
     **New** | **100000** |   **2** | **286.0879 us** |  **5.4825 us** |   **0.86** |     **-** |     **-** |   **254.91** |          **43,240.30** |
 OldPlus | 100000 |   2 | 330.7581 us |  5.6577 us |   1.00 |     - |     - | 1,098.28 |         169,005.43 |
     **New** | **100000** |  **10** | **371.8050 us** |  **3.3757 us** |   **0.92** |     **-** |     **-** |   **278.70** |          **64,468.67** |
 OldPlus | 100000 |  10 | 404.1597 us |  5.3421 us |   1.00 |     - |     - | 1,101.76 |         188,457.16 |
     **New** | **100000** |  **25** | **541.6640 us** | **10.5468 us** |   **0.92** |     **-** |     **-** |   **561.51** |          **88,952.07** |
 OldPlus | 100000 |  25 | 589.0407 us | 58.1685 us |   1.00 |     - |     - | 1,182.97 |         220,629.98 |
     **New** | **100000** |  **50** | **790.0072 us** | **13.8439 us** |   **1.00** |     **-** |     **-** | **1,322.00** |         **204,549.69** |
 OldPlus | 100000 |  50 | 787.2796 us |  6.5840 us |   1.00 |     - |     - | 1,576.00 |         244,716.71 |
     **New** | **100000** |  **75** | **656.1230 us** | **13.2718 us** |   **1.19** |     **-** |     **-** | **1,624.51** |         **254,655.45** |
 OldPlus | 100000 |  75 | 549.9804 us |  6.2456 us |   1.00 |     - |     - | 1,532.54 |         252,263.44 |
     **New** | **100000** |  **90** | **496.8645 us** | **10.6528 us** |   **1.01** |     **-** |     **-** | **1,956.62** |         **311,019.45** |
 OldPlus | 100000 |  90 | 490.2044 us | 40.9234 us |   1.00 |     - |     - | 1,775.84 |         315,199.39 |
     **New** | **100000** |  **98** | **481.1661 us** | **10.2857 us** |   **1.05** |     **-** |     **-** | **2,196.75** |         **327,805.23** |
 OldPlus | 100000 |  98 | 456.8940 us |  9.4714 us |   1.00 |     - |     - | 2,296.62 |         343,529.09 |
     **New** | **100000** | **100** | **474.2035 us** | **15.7986 us** |   **1.58** |     **-** |     **-** | **2,364.10** |         **357,264.53** |
 OldPlus | 100000 | 100 | 299.9004 us |  6.8316 us |   1.00 |     - |     - | 1,112.89 |         168,271.76 |

**Comment by KevinRansom** [comment]
The new algorithm for filter seems like the best choice.

Are there other places this techniques is employable ... is it generalizable?

Kevin

**Comment by jackmott** [comment]
Array.choose is the only place I know of where it is directly applicable.

**Comment by KevinRansom** [comment]
@jackmott 

Thanks for this contribution

Kevin

**Comment by manofstick** [comment]
@KevinRansom @jackmott 

Wow, I'm late to this party too!

Anyway, I tweeked by previous entry, and now it is often faster (and I think in real code, almost always faster) and except under extreme conditions always uses significantly less memory.

Go and [poke your head in at the gist](https://gist.github.com/manofstick/229384a1bd0bdb26caf9a780b952d9b8) to have a look. I have added a couple of parameters to tweek.

So first of all for random data:
- if the filter is returning just a few % or >~80% of the data it is faster.
- for small arrays (<1000 elements?) all filtering is faster.
- the worst case performance I have seen is < 15% slower than code currently in this PR. This was for an array size of 10000 and a filter of 50%. Further increases in size of the array decreased the difference.

For non-random ordered data, which I think is the bulk of data in the real world, as it often ends up being clustered in some fashion due to the way it is generated, or retrieved from a database, then the uses cases where the supplied version is better increases. I have modeled this in the gist with a parameter MaxRunLengthBeforeValueChange which is used as the max value of a random number used to populate the test array. If this value is >~50 then the alternative version in the gist always wins.

Anyway, here is the output from the gist:

     Method |   Length | MaxRunLengthBeforeValueChange | PercentOfDataInFilter |          Median |        StdDev | Scaled | Scaled-SD | Gen 0 | Gen 1 | Gen 2 | Bytes Allocated/Op |
----------- |--------- |------------------------------ |---------------------- |---------------- |-------------- |------- |---------- |------ |------ |------ |------------------- |
 Filter1456 |     1000 |                             1 |                     1 |       4.4961 us |     0.0833 us |   1.00 |      0.00 |  0.01 |     - |     - |             431.39 |
  Philter32 |     1000 |                             1 |                     1 |       4.0766 us |     0.1275 us |   0.91 |      0.03 |  0.00 |     - |     - |              96.96 |
 Filter1456 |     1000 |                             1 |                    10 |       5.8274 us |     0.1521 us |   1.00 |      0.00 |  0.02 |     - |     - |             734.54 |
  Philter32 |     1000 |                             1 |                    10 |       5.8186 us |     0.1548 us |   1.00 |      0.04 |  0.01 |     - |     - |             249.64 |
 Filter1456 |     1000 |                             1 |                    50 |      10.4041 us |     0.1634 us |   1.00 |      0.00 |  0.06 |     - |     - |           2,489.41 |
  Philter32 |     1000 |                             1 |                    50 |       8.3342 us |     0.2412 us |   0.80 |      0.03 |  0.03 |     - |     - |           1,076.26 |
 Filter1456 |     1000 |                             1 |                    90 |       6.8637 us |     0.1620 us |   1.00 |      0.00 |  0.09 |     - |     - |           3,821.80 |
  Philter32 |     1000 |                             1 |                    90 |       6.2562 us |     0.1150 us |   0.91 |      0.03 |  0.04 |     - |     - |           1,888.65 |
 Filter1456 |     1000 |                             1 |                    99 |       5.7584 us |     0.0974 us |   1.00 |      0.00 |  0.09 |     - |     - |           3,788.99 |
  Philter32 |     1000 |                             1 |                    99 |       5.4585 us |     0.0434 us |   0.95 |      0.02 |  0.05 |     - |     - |           2,252.80 |
 Filter1456 |     1000 |                           100 |                     1 |       3.9910 us |     0.1697 us |   1.00 |      0.00 |  0.00 |     - |     - |              48.02 |
  Philter32 |     1000 |                           100 |                     1 |       3.7195 us |     0.0897 us |   0.93 |      0.04 |  0.00 |     - |     - |              77.10 |
 Filter1456 |     1000 |                           100 |                    10 |       4.5371 us |     0.2308 us |   1.00 |      0.00 |  0.01 |     - |     - |             521.48 |
  Philter32 |     1000 |                           100 |                    10 |       3.9488 us |     0.1764 us |   0.89 |      0.06 |  0.01 |     - |     - |             255.63 |
 Filter1456 |     1000 |                           100 |                    50 |       5.1228 us |     0.1747 us |   1.00 |      0.00 |  0.07 |     - |     - |           2,537.71 |
  Philter32 |     1000 |                           100 |                    50 |       4.8066 us |     0.2096 us |   0.94 |      0.05 |  0.03 |     - |     - |           1,162.18 |
 Filter1456 |     1000 |                           100 |                    90 |       5.5019 us |     0.0624 us |   1.00 |      0.00 |  0.10 |     - |     - |           3,836.93 |
  Philter32 |     1000 |                           100 |                    90 |       5.3392 us |     0.2025 us |   0.96 |      0.04 |  0.05 |     - |     - |           2,021.98 |
 Filter1456 |     1000 |                           100 |                    99 |       5.5600 us |     0.0492 us |   1.00 |      0.00 |  0.10 |     - |     - |           4,115.52 |
  Philter32 |     1000 |                           100 |                    99 |       4.5134 us |     0.2643 us |   0.83 |      0.05 |  0.05 |     - |     - |           2,178.52 |
 Filter1456 | 10000000 |                             1 |                     1 |  45,149.4529 us |   431.8700 us |   1.00 |      0.00 |     - |     - |  0.42 |       7,605,202.92 |
  Philter32 | 10000000 |                             1 |                     1 |  43,166.1553 us |   103.9198 us |   0.95 |      0.01 |     - |     - |  0.45 |       4,845,177.50 |
 Filter1456 | 10000000 |                             1 |                    10 |  60,005.0910 us |   342.3094 us |   1.00 |      0.00 |     - |     - |  0.76 |      12,665,202.16 |
  Philter32 | 10000000 |                             1 |                    10 |  65,707.3230 us |   489.6155 us |   1.10 |      0.01 |     - |     - |  0.92 |      10,236,629.80 |
 Filter1456 | 10000000 |                             1 |                    50 | 110,770.2591 us |   940.7327 us |   1.00 |      0.00 |     - |     - | 27.42 |      23,253,907.10 |
  Philter32 | 10000000 |                             1 |                    50 | 122,336.2572 us |   125.6239 us |   1.10 |      0.01 |     - |     - |  0.92 |      15,725,496.51 |
 Filter1456 | 10000000 |                             1 |                    90 |  84,167.6096 us |   918.1276 us |   1.00 |      0.00 |     - |     - | 33.84 |      34,615,421.44 |
  Philter32 | 10000000 |                             1 |                    90 |  76,862.7960 us |   170.0396 us |   0.91 |      0.01 |     - |     - |  0.92 |      21,216,080.39 |
 Filter1456 | 10000000 |                             1 |                    99 |  74,072.9778 us |   826.2526 us |   1.00 |      0.00 |     - |     - | 28.68 |      32,228,365.12 |
  Philter32 | 10000000 |                             1 |                    99 |  60,613.1888 us |    51.1656 us |   0.82 |      0.01 |     - |     - |  0.92 |      23,253,517.65 |
 Filter1456 | 10000000 |                           100 |                     1 |  43,758.3237 us |   277.9877 us |   1.00 |      0.00 |     - |     - |  0.43 |       8,080,751.05 |
  Philter32 | 10000000 |                           100 |                     1 |  38,316.2922 us |    72.3523 us |   0.88 |      0.01 |     - |     - |  0.49 |       4,829,619.85 |
 Filter1456 | 10000000 |                           100 |                    10 |  46,055.5999 us |   129.8833 us |   1.00 |      0.00 |     - |     - |  0.38 |       8,376,540.38 |
  Philter32 | 10000000 |                           100 |                    10 |  43,129.5666 us | 1,230.1082 us |   0.95 |      0.03 |     - |     - |  0.45 |       6,261,245.66 |
 Filter1456 | 10000000 |                           100 |                    50 |  62,104.9744 us | 1,199.4773 us |   1.00 |      0.00 |     - |     - | 25.70 |      22,429,181.66 |
  Philter32 | 10000000 |                           100 |                    50 |  53,783.2780 us |   182.0459 us |   0.87 |      0.02 |     - |     - |  1.00 |      16,219,520.30 |
 Filter1456 | 10000000 |                           100 |                    90 |  70,464.7494 us |   867.5495 us |   1.00 |      0.00 |     - |     - | 28.20 |      30,510,707.10 |
  Philter32 | 10000000 |                           100 |                    90 |  58,599.4623 us |   373.3354 us |   0.83 |      0.01 |     - |     - |  0.81 |      19,861,687.28 |
 Filter1456 | 10000000 |                           100 |                    99 |  72,964.3383 us |   944.8472 us |   1.00 |      0.00 |     - |     - | 30.82 |      34,103,943.21 |
  Philter32 | 10000000 |                           100 |                    99 |  59,083.0047 us |   371.9605 us |   0.81 |      0.01 |     - |     - |  0.77 |      21,081,473.64 |

**Comment by jackmott** [comment]
@manofstick that tests really well for me. Sometimes slower, sometimes faster but allocations are way way down overall.

I wonder if we could tweak it even more. I really want to be able to just treat the bool as a number which would let us avoid the branches but don't know of any way to do that in F#.

Would 64 bits at a time be better with 64bit RyuJit?

**Comment by KevinRansom** [comment]
Durn it, that is pretty neat.

We have taken a 3 line method and turned it into a novel :-)

So we get an average 8.5 % performance improvement and 35% memory reduction if I did the sums correctly.

The approach is really interesting, It would never have occurred to me that we could have used a bit map and still got a performance improvement.  I suppose it comes from the loop unrolling and only having to write back to the bit element array 1 time for each 32 elements.

Although I do wonder if the additional code complexity is worth the gain.  Let me know what you think.  @jackmott had an interesting optimization that does very little work and doesn't allocate until after it has found a filtered value to return.  Which will almost certainly improve the perf and allocation size for the lower percent of data in filter cases. 
https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Core/array.fs#L490

Kevin

**Comment by jackmott** [comment]
It may be possible to combine that initial while loop optimization with this. I think it is worth playing with this approach a while, if we can refine it, either keep the same performance with simpler code, or extract more performance out of it, then it would be pretty compelling.

**Comment by jackmott** [comment]
Idea to try:
Do 64 at a time, then use a 64bit [hamming weight](http://stackoverflow.com/questions/2709430/count-number-of-bits-in-a-64-bit-long-big-integer) on `filterMask` (  to replace the `filteredCount <- filteredCount + 1`

**Comment by manofstick** [comment]
@jackmott 

OK, I'm not seriously proposing this, but was just doing this for a bit of fun re: your comment about bool to int conversion.

So you can [have a look here](https://gist.github.com/manofstick/0fed221b1ca4abcd65a632915d9dc48f) for some goodness and/or evilness depending on which hat your wearing.

Now funny thing with this code is that I find it quite a bit faster when just run in a standalone test (...where I still warm up and look and spin etc. etc...) but running under BenchmarketDotNet it is quite a bit slower... Someone with more time than I would need to investigate...

...Anyway, it's not particularly portable, so shouldn't be used, but just here for your amusement....

**Comment by jackmott** [comment]
@manofstick that is pretty awesome =)
I've played with your original one a bit and can't find any substantive improvements. I did try turning filteredMask into a 64bit uint, and then embedding the "+1" in it, which you can extract at the end with a mask or bit shift and a cast. But it was only faster in a few scenarios, and I'm not entirely confident it is a safe thing to do.

You should do a new PR with your filter, if you need help getting the solution to build let me know.

**Pull request merged**